[auth] Support auth levels for basic auth

gguillemas opened the pull request:
Thank you for submitting this pull request! We appreciate you spending the time to work on these changes.

## What is the motivation?

**Note**: This PR is a rebase of #2535, which will be closed as discussed with @sgirones. 

Before this PR, basic auth credentials were always validated against the `ROOT` level first. If the credentials were not valid for `ROOT`, the authentication process would try the same credentials for the `NS` provided on the `NS` header. If they weren't valid for `NS` either, the credentials would be tried for the `DB` provided on the `DB` header.

This caused problems when the `NS` or `DB` credentials matched the credentials for a `ROOT` user or when the `DB` credentials matched those for a `NS` user, because the credentials for the higher level user would always be used.

## What does this change do?

* For backward compatibility, this feature is enabled on the server with the `--enable-auth-level` flag on `start`.
* The `sql`, `export`, `import` and `backup` commands in the CLI now have the `auth-level` argument.
* The HTTP API now uses the `surreal-auth-ns` and `surreal-auth-db` headers when doing basic authentication to determine at which level the credentials need to be validated.
* All SurrealDB headers now use the `surreal-` prefix while still supporting legacy headers without the prefix.

### To Do

- [x] Add new tests.
- [x] Decide the headers suffix. Do we use `sur-`?
- [x] Decide whether or not to be backwards compatible at this point for the headers.

Taking advantage of this optional feature in the SDKs will require supporting the `auth_level` option to let clients specify the desired authentication level (i.e. `root`, `namespace`, `database` and aliases) and adding the necessary HTTP headers (i.e. `surreal-auth-ns` and `surreal-auth-db`) in order to select the namespace and/or database that the authenticating user belongs to. Since the default behaviour is backwards compatible, this PR can be merged and released before updating the SDKs without breaking compatibility.  The expected behaviour can be seen in the HTTP integration tests.

## What is your testing strategy?

Added several `http_integration` and `cli_integration` tests to validate both new and legacy behavior.

## Is this related to any issues?

No.

## Have you read the [Contributing Guidelines](https://github.com/surrealdb/surrealdb/blob/main/CONTRIBUTING.md)?

- [x] I have read the [Contributing Guidelines](https://github.com/surrealdb/surrealdb/blob/main/CONTRIBUTING.md)

kearfy commented:
![image](https://github.com/surrealdb/surrealdb/assets/17960250/57307ee1-7d04-4485-95ac-0cdd22651e16)

gguillemas commented:
Implemented the suggestions made by @rushmorem in #2535.
- [d9a9a37](https://github.com/surrealdb/surrealdb/pull/2985/commits/d9a9a37bd4726483a148dc7f34f6c6712ef96eb6): Changed HTTP header prefix from `sur-` to `surreal-`. E.g. `surreal-ns`, `surreal-db`.
- [4061c2e](https://github.com/surrealdb/surrealdb/pull/2985/commits/4061c2ee427b4a4458ca8b7f128b9200c3beb438): Changed `with_something` methods to expect `impl Into<Option<&'a str>>` so they can accept string literals.
- [6b31f16](https://github.com/surrealdb/surrealdb/pull/2985/commits/6b31f166e43d7ef826c1ef267353e32cf44177cd): Removed the prefix from accessor methods. E.g. from `creds.from_root()` to simply `creds.root()`.

rushmorem commented on lib/src/api/opt/endpoint/mod.rs:
```diff
@@ -92,3 +102,49 @@ mod tests {
+
+impl EndpointKind {
+	pub fn is_local(&self) -> bool {
+		!matches!(
+			self,
+			EndpointKind::Http | EndpointKind::Https | EndpointKind::Ws | EndpointKind::Wss
+		)
```
I know we have discussed this before @sgirones but this still makes me uneasy. How about re-implimenting this method as `is_remote`. We could then re-add `is_local` as

```rust
pub fn is_local(&self) -> bool {
    !self.is_remote()
}
```

if we like.

gguillemas replied:
I agree that it looks better, changed in [f49e028](https://github.com/surrealdb/surrealdb/pull/2985/commits/f49e0287c5fe64c46a731de1f264e0768c6823f1).

rushmorem commented on src/err/mod.rs:
```diff
@@ -34,6 +36,9 @@ pub enum Error {
 	InvalidStorage,
 
+	#[error("There was a problem parsing the header {0}: {1}")]
+	InvalidHeader(HeaderName, TypedHeaderRejection),
```
`axum` hasn't reached 1.0 yet. Making it a public dependency like this in a stable crate will be problematic.

rushmorem commented on lib/src/lib.rs:
```diff
@@ -139,6 +139,9 @@ pub mod syn;
 pub use api::engine;
+#[cfg(feature = "protocol-http")]
+#[doc(inline)]
+pub use api::headers;
```
Do we expect users to use these headers or we are only exporting them for the server?

gguillemas replied:
As we discussed, I have kept the header constants out of the public API in [3aa88f7](https://github.com/surrealdb/surrealdb/pull/2985/commits/3aa88f7cfd52825d06d4789f0fd919bfe740f359).

rushmorem commented on lib/src/api/opt/endpoint/mod.rs:
```diff
@@ -92,3 +102,49 @@ mod tests {
 }
+
+#[derive(Debug)]
+pub enum EndpointKind {
```
Let's also avoid making this part of the public API for now.

gguillemas replied:
As we discussed, I have kept `EndpointKind` out of the public API in [3aa88f7](https://github.com/surrealdb/surrealdb/pull/2985/commits/3aa88f7cfd52825d06d4789f0fd919bfe740f359).

sgirones commented on lib/src/iam/verify.rs:
```diff
@@ -86,11 +86,55 @@ pub async fn basic(
+	}
+}
+
+pub async fn basic_legacy(
```
```suggestion
# TODO(gguillemas): Remove this method once the legacy basic auth is deprecated in v2.0.0
pub async fn basic_legacy(
```

sgirones commented on lib/src/iam/verify.rs:
```diff
@@ -482,6 +500,47 @@ fn verify_pass(pass: &str, hash: &str) -> Result<(), Error> {
 	}
 }
 
+pub async fn verify_creds_legacy(
```
```suggestion
# TODO(gguillemas): Remove this method once the legacy basic auth is deprecated in v2.0.0
pub async fn verify_creds_legacy(
```

sgirones commented on tests/http_integration.rs:
```diff
@@ -7,20 +7,219 @@ mod http_integration {
+	}
+
+	#[test(tokio::test)]
+	async fn basic_auth_legacy() -> Result<(), Box<dyn std::error::Error>> {
```
```suggestion
        # TODO(gguillemas): Remove this test once the legacy basic auth is deprecated in v2.0.0
	async fn basic_auth_legacy() -> Result<(), Box<dyn std::error::Error>> {
```

sgirones commented on src/cli/start.rs:
```diff
@@ -74,6 +74,12 @@ pub struct StartCommandArguments {
 		requires = "username"
 	)]
 	password: Option<String>,
+	#[arg(
```
```suggestion
        # TODO(gguillemas): Remove this arg once the legacy basic auth is deprecated in v2.0.0, because it will be enabled by default
	#[arg(
```

gguillemas commented on lib/src/api/opt/endpoint/mod.rs:
```diff
@@ -33,6 +34,15 @@ pub struct Endpoint {
 }
 
+impl Endpoint {
+	pub fn parse_kind(&self) -> Result<EndpointKind> {
```
As we discussed, I have kept `EndpointKind` out of the public API in [3aa88f7](https://github.com/surrealdb/surrealdb/pull/2985/commits/3aa88f7cfd52825d06d4789f0fd919bfe740f359).

gguillemas commented:
The changes that we agreed regarding maintaining backward compatibility in the behavior of basic authentication have been implemented together with the suggestions made in the code reviews. This PR is again ready to be reviewed!

rushmorem commented on lib/src/api/opt/auth.rs:
```diff
@@ -12,6 +13,108 @@ pub struct Signup;
 
+/// Credentials level
+#[derive(Debug, Clone)]
+pub enum CredentialsLevel {
```
Does this need to be part of the API? Are exposing it anywhere in the SDK? If not, can we hide it too?

gguillemas replied:
Thank you! I have moved these types to the CLI in [526a849](https://github.com/surrealdb/surrealdb/pull/2985/commits/526a849fa33e5194ac207ffc69d2ead389e98c28).

rushmorem commented on lib/src/api/opt/auth.rs:
```diff
@@ -12,6 +13,108 @@ pub struct Signup;
+
+/// Construct a Credentials instance for the given auth level
+#[derive(Debug, Default)]
+pub struct CredentialsBuilder<'a> {
```
All these new additions in this module make sense to me from the CLI not the SDK. Is there any reason these are in the SDK? For example, I don't think using a builder here is an improvement to using the structs directly like we already do. On the contrary, it now makes it possible to make mistakes that were not possible before, hence we now need to introduce a new error type.

gguillemas replied:
Thank you! I have moved these types to the CLI in [526a849](https://github.com/surrealdb/surrealdb/pull/2985/commits/526a849fa33e5194ac207ffc69d2ead389e98c28).
